posts/2024-05-07-ultimate-string-guide.md

```diff
@@ -0,0 +1,625 @@
+
+The upside of this is that it's simple, the downside is that it's wasting space, because most values don't
+need the whole 21 bits (e.g. ASCII just needs 7 bits).
+
+UTF-32 is not ASCII compatible, meaning a program that only understands ASCII won't accidentially work with UTF-32 text,
+even if all of the characters used are in the ASCII set (e.g. only latin characters from `[a-zA-Z]`).
+
+### UTF-16
+
+This is a [variable-width character encoding](https://en.wikipedia.org/wiki/Variable-width_encoding), most notably
+[used on windows](https://learn.microsoft.com/en-us/windows/win32/learnwin32/working-with-strings).
+
+Since we only have access to 16 bits per machine word, only the code points from `U+0000` to `U+FFFF`
+(with the exception of surrogates, which I will explain later) are encoded "directly".
+
+Code points from `U+10000` to `U+10FFFF` need two 16-bit words. In order to encode these without
+being accidentially ambiguous, **surrogates** were introduced (another option would have been magic bits as used by UTF-8,
+but I guess the format wasn't designed with extension in mind). These surrogates must always come in pairs and express:
+
+- low surrogates: `U+DC00` to `U+DFFF`
+- high surrogates: `U+D800` to `U+DBFF`
+
+Through bit shuffling, these two bytes allows to map to values in the `U+10000` to `U+10FFFF` range.
+For the interested reader, the algorithm is as follows
+([cited from wikipedia](https://en.wikipedia.org/wiki/UTF-16#Code_points_from_U+010000_to_U+10FFFF)):
+
+> * 0x10000 is subtracted from the code point (U), leaving a 20-bit number (U') in the hex number range 0x00000–0xFFFFF.
+> * The high ten bits (in the range 0x000–0x3FF) are added to 0xD800 to give the first 16-bit code unit or high surrogate (W1), which will be in the range 0xD800–0xDBFF.
+> * The low ten bits (also in the range 0x000–0x3FF) are added to 0xDC00 to give the second 16-bit code unit or low surrogate (W2), which will be in the range 0xDC00–0xDFFF
+
+UTF-16 is not ASCII compatible either. It is more space efficient than UTF-32 though. For some languages, it can
+even be more space efficient than UTF-8.
+
+
+### Unicode Scalar Values
+
+It is important to understand that the Haskell `Char` type (which is essentially a Code Point) can represent surrogates
+that are used in UTF-16.
+
+The unicode standard also defines the concept of [Unicode Scalar Values](http://www.unicode.org/versions/Unicode5.2.0/ch03.pdf#page=35):
+
+> Any Unicode code point except high-surrogate and low-surrogate code points. In other words, the ranges of integers 0 to D7FF16 and E00016 to 10FFFF16 inclusive.
+
+So, code point without surrogates. This will become relevant for UTF-8.
+
+### UTF-8
+
+This is similar to UTF-16 a variable-width character encoding. It's often used in web APIs (most notably JSON) and is
+often the default on unix systems.
+
+Here a word is only 8 bit. The number of bytes required depends on the range of the code point and
```

Comment:
"word"?

Reply:
It's actually "code unit", but I found it confusing, so was looking for a simpler expression.